Remove legacy `Parse(U)Int*`

maflcko opened the pull request:
The legacy int parsing is problematic, because it accepts the `+` sign for unsigned integers. In all cases this is either:

* Useless, because the `+` sign was already rejected.
* Erroneous and inconsistent, when third party parsers reject it. (C.f. https://github.com/bitcoin/bitcoin/pull/32365)
* Confusing, because the `+` sign is  neither documented, nor can it be assumed to be present.

Fix all issues by removing the legacy int parsing.

Commit:
refactor: Use ToIntegral in ParseHDKeypath

ToIntegral<uint32_t> only accepts numbers, so just use that to replace
the equivalent but more verbose way with find_first_not_of+ParseUInt32.

Commit:
refactor: Use ToIntegral in CreateFromDump

The unsigned version is never serialized with a `+` prefix, so the
stricter ToIntegral can be used.

Commit:
cli: Reject + sign in -netinfo level parsing

It would be confusing to specify the sign for an unsigned value here, so
reject it.

Commit:
bitcoin-tx: Reject + sign in nversion parsing

It would be confusing to specify the sign for an unsigned value here, so
reject it.

Commit:
init: Reject + sign in -*port parsing

It would be confusing to specify the sign for an unsigned value here, so
reject it.

Commit:
net: Reject + sign when parsing subnet mask

It does not make sense and it is rejected by other parsers as well:

>>> ipaddress.ip_network("1.2.3.0/+24")
ValueError: '1.2.3.0/+24' does not appear to be an IPv4 or IPv6 network

Commit:
Reject + sign in SplitHostPort

It is better to reject it with an error. For example,

$ bitcoin-cli -rpcconnect=127.0.0.1:+23501 -getinfo
error: Invalid port provided in -rpcconnect: 127.0.0.1:+23501

Commit:
Reject + sign when checking -ipcfd

Commit:
test: Reject + sign when parsing regtest deployment params

The + sign does not make sense for times or heights.

Commit:
rest: Reject + sign in /blockhashbyheight/

Commit:
bitcoin-tx: Reject + sign in locktime

Commit:
bitcoin-tx: Reject + sign in MutateTxDel*

Commit:
bitcoin-tx: Reject + sign in replaceable parsing

Commit:
bitcoin-tx: Reject + sign in vout parsing

Commit:
refactor: Remove unused Parse(U)Int*

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32520/commits/fa24012c181850d4adc5cca0fdba76d584ec3a00

brunoerg reviewed (commented):
Concept ACK

stickies-v commented on src/bitcoin-tx.cpp:
```diff
@@ -229,16 +229,15 @@ static void MutateTxLocktime(CMutableTransaction& tx, const std::string& cmdVal)
-    // parse requested index
-    int64_t inIdx = -1;
-    if (strInIdx != "" && (!ParseInt64(strInIdx, &inIdx) || inIdx < 0 || inIdx >= static_cast<int64_t>(tx.vin.size()))) {
+    const auto idx{ToIntegral<int64_t>(strInIdx)};
```
nit: I think making this a `size_t` simplifies it a bit more?

<details>
<summary>git diff on fa24012c18</summary>

```diff
diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index 816f11045a..9e2c6d0902 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -229,13 +229,13 @@ static void MutateTxLocktime(CMutableTransaction& tx, const std::string& cmdVal)
 
 static void MutateTxRBFOptIn(CMutableTransaction& tx, const std::string& strInIdx)
 {
-    const auto idx{ToIntegral<int64_t>(strInIdx)};
-    if (strInIdx != "" && (!idx || *idx < 0 || *idx >= static_cast<int64_t>(tx.vin.size()))) {
+    const auto idx{ToIntegral<size_t>(strInIdx)};
+    if (strInIdx != "" && (!idx  || *idx >= tx.vin.size())) {
         throw std::runtime_error("Invalid TX input index '" + strInIdx + "'");
     }
 
     // set the nSequence to MAX_INT - 2 (= RBF opt in flag)
-    int cnt = 0;
+    size_t cnt = 0;
     for (CTxIn& txin : tx.vin) {
         if (strInIdx == "" || cnt == *idx) {
             if (txin.nSequence > MAX_BIP125_RBF_SEQUENCE) {

```
</details>

stickies-v replied:
Ah, good point, that seems sensible. (+ for [other](https://github.com/bitcoin/bitcoin/pull/32520#discussion_r2093431546))

stickies-v commented on src/bitcoin-tx.cpp:
```diff
@@ -507,26 +507,20 @@ static void MutateTxAddOutScript(CMutableTransaction& tx, const std::string& str
-    // parse requested deletion index
-    int64_t inIdx;
-    if (!ParseInt64(strInIdx, &inIdx) || inIdx < 0 || inIdx >= static_cast<int64_t>(tx.vin.size())) {
+    const auto idx{ToIntegral<uint64_t>(strInIdx)};
```
nit: looks safe because we're comparing with `tx.vin.size()`, but using `size_t` seems more appropriate for a container index? (+ for `MutateTxDelOutput`)

stickies-v commented on src/test/fuzz/parse_numbers.cpp:
```diff
@@ -14,24 +14,9 @@ FUZZ_TARGET(parse_numbers)
     (void)ParseMoney(random_string);
 
-    uint8_t u8;
-    (void)ParseUInt8(random_string, &u8);
```
It seems we now no longer have `ToIntegral` fuzz test coverage. It's a fairly simple function, so maybe that's fine, but so was `ParseIntegral`. Reckon it's worth keeping `ToIntegral` for the most common types in this harness?

stickies-v reviewed (approved):
ACK fa24012c181850d4adc5cca0fdba76d584ec3a00, nice cleanup.

Introduces some slight behaviour change when passing integer arguments through startup options, the REST interface, as and through `bitcoin-cli` and `bitcoin-tx`. While some of these changes _may_ break existing user flows that were previously fine, they should hopefully be quite rare, and I think are worth the cleanup and more strict parsing.

Commit:
rest: Use SAFE_CHARS_URI in SanitizeString error msg

When dealing with URI parts, it seems more consistent to use
corresponding SAFE_CHARS_URI mode in error messages.

Co-Authored-By: stickies-v <stickies-v@protonmail.com>

stickies-v reviewed (approved):
re-ACK fa2445eea3e380386ab01631b735c7eeb6b76b58

stickies-v commented:
re-ACK 33332829333b589420f8038541d04ec6970f051d

fjahr commented:
utACK 33332829333b589420f8038541d04ec6970f051d

I think I found two leftover mentions of the `ParseInt` in documentation: https://github.com/fjahr/bitcoin/commit/1f2f1a432b0fdb3d84eecc9dc3a2db34d88dc892 I can option a follow-up if you don't want to address it here.

Commit:
doc: Remove ParseInt mentions in documentation

In the dev notes, remove the whole section, because:

* ParseDouble was removed in commit
  fa9d72a7947d2cff541794e21e0040c3c1d43b32
* The locale-dependent atoi is already checked by
  test/lint/lint-locale-dependence.py

Co-authored-by: Fabian Jahr <fjahr@protonmail.com>

stickies-v commented:
re-ACK faf55fc80b

brunoerg reviewed (approved):
code review ACK faf55fc80b11f3d9b0b12c1b26a9612ea9ce9b40

Did a grep to check if anything is missing, seems fine.
